File: android-native/app/src/main/java/org/tvplayer/app/Channel.java
package org.tvplayer.app;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

public class Channel {
    public final String name;
    public final String group;
    public final String key;
    private final List<String> urls;

    public Channel(String name, String url, String group) {
        this(name, group, M3UParser.normalizeName(name), new ArrayList<String>());
        addUrl(url);
    }

    public Channel(String name, String group, String key, List<String> urls) {
        this.name = name != null && !name.trim().isEmpty() ? name.trim() : "未知";
        this.group = group != null && !group.trim().isEmpty() ? group.trim() : "未分组";
        this.key = key != null && !key.trim().isEmpty() ? key.trim() : M3UParser.normalizeName(this.name);
        this.urls = new ArrayList<>();
        if (urls != null) {
            for (String url : urls) {
                addUrl(url);
            }
        }
    }

    public void addUrl(String url) {
        if (url == null) {
            return;
        }
        String clean = url.trim();
        if (clean.isEmpty() || urls.contains(clean)) {
            return;
        }
        urls.add(clean);
    }

    public List<String> getUrls() {
        return Collections.unmodifiableList(urls);
    }

    public String getPrimaryUrl() {
        return urls.isEmpty() ? "" : urls.get(0);
    }

    public int getSourceCount() {
        return urls.size();
    }

    public String getStorageKey() {
        return key;
    }
}

File: TVPlayer-iOS/TVPlayer_iOS/Model/Channel.swift
import Foundation

/// 频道数据模型 — 值类型，线程安全，Codable
struct Channel: Codable, Identifiable, Equatable, Hashable {
    /// 稳定 id：用 key 避免列表 diff 闪烁
    var id: String { key }

    let name: String
    let group: String
    let key: String
    private(set) var urls: [String]

    enum CodingKeys: String, CodingKey {
        case name, group, key, urls
    }

    init(name: String, group: String = "未分组", key: String? = nil, urls: [String]? = nil) {
        let trimmedName = name.trimmingCharacters(in: .whitespaces)
        self.name = trimmedName.isEmpty ? "未知" : trimmedName
        let trimmedGroup = group.trimmingCharacters(in: .whitespaces)
        self.group = trimmedGroup.isEmpty ? "未分组" : trimmedGroup
        self.key = (key?.trimmingCharacters(in: .whitespaces)).flatMap { $0.isEmpty ? nil : $0 }
            ?? M3UParserService.normalizeName(trimmedName)
        self.urls = []
        if let urls {
            for u in urls {
                addUrl(u)
            }
        }
    }

    /// 添加 URL（去重 + 清理）
    mutating func addUrl(_ url: String) {
        let clean = url.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !clean.isEmpty, !urls.contains(clean) else { return }
        urls.append(clean)
    }

    /// 批量添加 URL
    mutating func addUrls(_ newUrls: [String]) {
        for u in newUrls {
            addUrl(u)
        }
    }

    /// 删除指定 URL
    mutating func removeUrl(_ url: String) {
        let clean = url.trimmingCharacters(in: .whitespacesAndNewlines)
        urls.removeAll { $0 == clean }
    }

    var sourceCount: Int { urls.count }

    var primaryUrl: String { urls.first ?? "" }

    var isEmpty: Bool { urls.isEmpty }

    // MARK: - Hashable

    func hash(into hasher: inout Hasher) {
        hasher.combine(key)
    }

    // MARK: - Equatable

    static func == (lhs: Channel, rhs: Channel) -> Bool {
        lhs.key == rhs.key
    }

    /// 合并两个同名频道的 URL
    mutating func merge(with other: Channel) {
        for url in other.urls {
            addUrl(url)
        }
    }
}
